# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) More SQL
Week 5 | Lesson 4.1

### LEARNING OBJECTIVES
*After this lesson, you will be able to:*
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases (`AS`)
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`


### STUDENT PRE-WORK
*Before this lesson, you should already be able to:*
- connect to a local or remote relational database
- add and remove data, create new tables, alter table schemas
- perform simple queries using SQL language

### LESSON GUIDE
| TIMING  | TYPE  | TOPIC  |
|:-:|---|---|
| 5 mins | [Opening](#opening) | Opening |
| 10 mins | [Introduction](#introduction) | More SQL |
| 10 mins | [Guided-practice](#guided-practice_1) | Guided Practice: ORDER BY |
| 10 mins | [Guided-practice](#guided-practice_2) | Guided Practice: Aliases |
| 10 mins | [Guided-practice](#guided-practice_3) | Guided Practice: SQL LIKE Operator |
| 10 mins | [Guided-practice](#guided-practice_4) | Guided Practice: DISTINCT and LIMIT Operators |
| 10 mins | [Guided-practice](#guided-practice_5) | Guided Practice: GROUP BY Operator |
| 10 mins | [Guided-practice](#guided-practice_6) | Guided Practice: HAVING Operator |
| 10 mins | [Ind-practice](#ind-practice) | Independent Practice |
| 5 mins | [Conclusion](#conclusion) | Conclusion |

<a name="opening"></a>
## Opening (5 mins)
We have seen how to connect to a local sqlite database and to a remote postgresql database.

> **Check:** What SQL commands have we learned so far? What commands have we learned that are specific to SQLite and PostgreSQL?




> Answer:
- CREATE
- INSERT
- DELETE
- UPDATE
- SELECT
- how to list schema and tables

<a name="introduction"></a>
## More SQL (10 mins)

In this lecture we'll learn a few more useful `SQL` commands that will give us the ability to perform more complex queries.

- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`




For all the following example we will use the well-known [Northwind sample database](https://northwinddatabase.codeplex.com/).

We have imported this database into our PostgreSQL instance that you can find connecting here:

    psql -h dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com -p 5432 -U dsi_student northwind
    password: gastudents

We will use a few tables from this database, here is what they look like:



In [24]:
import psycopg2
from pandas.io import sql
login={ 'dbname': 'northwind',
      'user':'dsi_student',
      'port':5432,
      'host':'dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com',
      'password':'gastudents'}

conn=psycopg2.connect(**login)
c=conn.cursor()

In [30]:
sql.read_sql("SELECT * FROM customers LIMIT 3;",con=conn)

CustomerID                         CompanyName     ContactName  \
0      ALFKI                 Alfreds Futterkiste    Maria Anders   
1      ANATR  Ana Trujillo Emparedados y helados    Ana Trujillo   
2      ANTON             Antonio Moreno Taquería  Antonio Moreno   

           ContactTitle                        Address         City Region  \
0  Sales Representative                  Obere Str. 57       Berlin   None   
1                 Owner  Avda. de la Constitución 2222  México D.F.   None   
2                 Owner                Mataderos  2312  México D.F.   None   

  PostalCode  Country         Phone           Fax  
0      12209  Germany   030-0074321   030-0076545  
1      05021   Mexico  (5) 555-4729  (5) 555-3745  
2      05023   Mexico  (5) 555-3932          None

In [33]:
sql.read_sql("SELECT \"ContactTitle\" FROM customers LIMIT 5;",con=conn)

ContactTitle
0  Sales Representative
1                 Owner
2                 Owner
3  Sales Representative
4   Order Administrator

`customers`:

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|ALFKI| Alfreds Futterkiste| Maria Anders | Sales Representative | Obere Str. 57 | Berlin|| 12209| Germany | 030-0074321| 030-0076545|
|ANATR| Ana Trujillo Emparedados y helados | Ana Trujillo | Owner| Avda. de la Constitución 2222 | México D.F. || 05021| Mexico| (5) 555-4729 | (5) 555-3745|
|ANTON| Antonio Moreno Taquería| Antonio Moreno | Owner| Mataderos2312 | México D.F. || 05023| Mexico| (5) 555-3932 |
|...|...|...|...|...|...|...|...|...|...|...|



`orders`:

|OrderID | CustomerID | EmployeeID | OrderDate| RequiredDate | ShippedDate | ShipVia | Freight | ShipName|ShipAddress |ShipCity| ShipRegion | ShipPostalCode | ShipCountry |
|----|
|10248 | VINET|5 | 1996-07-04 | 1996-08-01 | 1996-07-16| 3 | 32.38 | Vins et alcools Chevalier | 59 rue de l'Abbaye | Reims|| 51100| France|
|10249 | TOMSP|6 | 1996-07-05 | 1996-08-16 | 1996-07-10| 1 | 11.61 | Toms Spezialitäten| Luisenstr. 48| Münster|| 44087| Germany|
|10250 | HANAR|4 | 1996-07-08 | 1996-08-05 | 1996-07-12| 2 | 65.83 | Hanari Carnes | Rua do Paço, 67| Rio de Janeiro | RJ | 05454-876| Brazil|
|...|...|...|...|...|...|...|...|...|...|...|



In [ ]:
sql.

`order_details`:

| OrderID |  ProductID |  UnitPrice | Quantity | Discount |
| ----- |
|10248|11|14|12|0|
|10248|42|9.8|10|0|
|10248|72|34.8|5|0|
|10249|14|18.6|9|0|
|10249|51|42.4|40|0|
|10250|41|7.7|10|0|
|...|...|...|...|...|


> **Check** Let's come up with a few queries you'd be curious to run on these tables.

### An ERD for our database

![](./assets/images/Northwind_A4_size_for_Print.png) 

<a name="guided-practice_1"></a>
## Guided Practice: ORDER BY (10 mins)

The `ORDER BY` keyword is used to sort the result-set by one or more columns. It sorts the records in ascending order by default. To sort the records in a descending order, you can use the `DESC` keyword.

### SQL ORDER BY Syntax

```sql
SELECT _column_name_,_ column_name_  
FROM _table_name_  
ORDER BY _column_name _ASC|DESC,_ column_name_ ASC|DESC;
```

#### Example

The following SQL statement selects all customers from the "Customers" table,  sorted by the "Country" column:

```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country";
```

|CustomerID|CompanyName|Country|
|----|
|OCEAN|Océano Atlántico Ltda.|Argentina|
|CACTU|Cactus Comidas para llevar|Argentina|
|RANCH|Rancho grande |Argentina|
|...|...|...|


#### ORDER BY DESC Example

The following SQL statement selects all customers from the "Customers" table,  sorted DESCENDING by the "Country" column:


```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country" DESC;
```

|CustomerID|CompanyName|Country|
|----|
|GROSR|GROSELLA-Restaurante |Venezuela|
|LILAS|LILA-Supermercado|Venezuela|
|HILAA|HILARION-Abastos |Venezuela|
|...|...|...|



#### ORDER BY Several Columns Example

The following SQL statement selects all customers from the `customers` table,  sorted by the "Country" and the "CustomerName" column:

```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country", "CompanyName";
```

|CustomerID|CompanyName|Country|
|----|
|CACTU|Cactus Comidas para llevar|Argentina|
|OCEAN|Océano Atlántico Ltda.|Argentina|
|RANCH|Rancho grande |Argentina|
|...|...|...|


Note that in `PostgreSQL` unquoted names are case-insensitive. Thus `SELECT * FROM hello` and `SELECT * FROM HELLO` are equivalent. (Try it.)

However, quoted names are case-sensitive. `SELECT * FROM "hello"` is not equivalent to `SELECT * FROM "HELLO"`.

To make a "bridge" between quoted names and unquoted names, unquoted names are implicitly lowercased, thus hello, HELLO and HeLLo are equivalent to "hello", but not to "HELLO" or "HeLLo".

Thus, when creating entities (tables, views, procedures, etc) in PostgreSQL, you should specify them either unquoted, or quoted-but-lowercased.

> **Check** run a few queries on the provided database and try using the `ORDER BY` command.

In [37]:
sql.read_sql("SELECT \"CustomerID\", \"CompanyName\", \"Country\" FROM customers ORDER BY \"Country\" DESC, \"CompanyName\" LIMIT 5;",con=conn)

CustomerID              CompanyName    Country
0      GROSR     GROSELLA-Restaurante  Venezuela
1      HILAA         HILARION-Abastos  Venezuela
2      LILAS        LILA-Supermercado  Venezuela
3      LINOD         LINO-Delicateses  Venezuela
4      GREAL  Great Lakes Food Market        USA

<a name="guided-practice_2"></a>
## Guided Practice: Aliases (10 mins)

SQL aliases are used to give a database table, or a column in a table, a temporary name. These can make column names more readable, or assign results of a query to an alias for further operations.

### SQL Alias Syntax for Columns

```sql
SELECT _column_name_ AS _alias_name_  
FROM _table_name_;
```
### SQL Alias Syntax for Tables
```sql
SELECT _alias_name_._column_name(s)_  
FROM _table_name _AS _alias_name_;
```

#### Alias Examples for Table Columns

The following SQL statement specifies two aliases, one for the `CompanyName`  column and one for the `ContactName` column:

```sql
SELECT "CompanyName" AS "Customer", "ContactName" AS "[Contact Person]"
FROM customers;
```

|Customer|[Contact Person]|
|----|
|Alfreds Futterkiste| Maria Anders|
|Ana Trujillo Emparedados y helados| Ana Trujillo|
|Antonio Moreno Taquería| Antonio Moreno|
|...|...|


In the following SQL statement we combine four columns (Address, City, PostalCode and Country) and create an alias named `Address`:

```sql
SELECT "CompanyName" AS "Customer",
    CONCAT("Address",', ',"City",', ', "PostalCode",', ',"Country") AS "Address"
FROM customers;
```

|Customer|Address|
|----|
|Alfreds Futterkiste| Obere Str. 57, Berlin, 12209, Germany|
|Ana Trujillo Emparedados y helados| Avda. de la Constitución 2222, México D.F., 05021, Mexico|
|Antonio Moreno Taquería Mataderos| 2312, México D.F., 05023, Mexico|
|...|...|


#### Alias Example for Tables

The following SQL statement selects all the orders from the customer with  CustomerID=4 (Around the Horn). We use the `customers` and `orders` tables, and  give them the table aliases of `c` and `o` respectively (Here we have used  aliases to make the SQL shorter):


```sql
SELECT o."OrderID", o."OrderDate", c."CompanyName"
FROM customers AS c, orders  AS o  
WHERE c."CompanyName" = 'Around the Horn' AND c."CustomerID"=o."CustomerID";
```

|CustomerID|OrderDate|CompanyName|
|----|
|10355|1996-11-15|Around the Horn|
|10383|1996-12-16|Around the Horn|
|10453|1997-02-21|Around the Horn|
|10558|1997-06-04|Around the Horn|
|...|...|...|

> **Check** what would be the same SQL statement without aliases?

Aliases can be useful when:

- There are more than one table involved in a query
- Functions are used in the query
- Column names are big or not very readable
- Two or more columns are combined together

> **Check** try running some of your earlier queries with aliases. Are they more readable?

In [46]:
sql.read_sql("SELECT \"CompanyName\" AS \"Customer\", \
    CONCAT(\"Address\",', ',\"City\",', ', \"PostalCode\",', ',\"Country\") AS \"Address\" \
FROM customers WHERE \"CompanyName\" LIKE \'%Market%\' ORDER BY \"CompanyName\" Desc;",con=conn)

Customer                                           Address
0     White Clover Markets  305 - 14th Ave. S. Suite 3B, Seattle, 98128, USA
1       Save-a-lot Markets                187 Suffolk Ln., Boise, 83720, USA
2  Great Lakes Food Market              2732 Baker Blvd., Eugene, 97403, USA
3    Bottom-Dollar Markets    23 Tsawassen Blvd., Tsawassen, T2F 8M4, Canada

<a name="guided-practice_3"></a>
## Guided Practice: SQL LIKE Operator (10 mins)

The LIKE operator is used in a WHERE clause to search for a specified pattern in a column.

### SQL LIKE Syntax
```sql
SELECT _column_name(s)_  
FROM _table_name_  
WHERE _column_name_ LIKE _pattern_;
```

#### SQL LIKE Operator Examples

The following SQL statement selects all customers with a City starting with  the letter "S":

```sql
SELECT * FROM customers
WHERE "City" LIKE 'S%';
```

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|BLONP| Blondesddsl père et fils|Frédérique Citeaux|Marketing Manager| 24, place Kléber|Strasbourg||67000| France|88.60.15.31|88.60.15.32|
|COMMI| Comércio Mineiro|Pedro Afonso|Sales Associate|Av. dos Lusíadas, 23|Sao Paulo| SP|05432-043| Brazil|(11) 555-7647| 
|FAMIA| Familia Arquibaldo|Aria Cruz| Marketing Assistant | Rua Orós, 92|Sao Paulo| SP|05442-030| Brazil|(11) 555-9857| 
|...|...|...|...|...|...|...|...|...|...|...|

**Tip:** The "%" sign is used to define wildcards (missing  letters) both before and after the pattern. Also notice the use of single quotation marks, and the case sensitivity.

The following SQL statement selects all customers with a City ending with the  letter "s":

```sql
SELECT * FROM customers
WHERE "City" LIKE '%s';
```

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|CACTU| Cactus Comidas para llevar|Patricio Simpson|Sales Agent|Cerrito 333 |Buenos Aires||1010|Argentina| (1) 135-5555|(1) 135-4892|
|DUMON| Du monde entier|Janine Labrune|Owner| 67, rue des Cinquante Otages|Nantes||44000| France|40.67.88.88| 40.67.89.89|
|FRANR| France restauration|Carine Schmitt|Marketing Manager| 54, rue Royale|Nantes||44000| France|40.32.21.21| 40.32.21.20|
|...|...|...|...|...|...|...|...|...|...|...|

Using the NOT keyword allows you to select records that do NOT match the pattern. The following SQL statement selects all customers with Country NOT  containing the pattern "land":

```sql
SELECT * FROM customers
WHERE "Country" NOT LIKE '%land%';
```
|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|ALFKI| Alfreds Futterkiste|Maria Anders|Sales Representative|Obere Str. 57| Berlin||12209| Germany|030-0074321| 030-0076545|
|ANATR| Ana Trujillo Emparedados y helados|Ana Trujillo|Owner| Avda. de la Constitución 2222| México D.F.|| 05021| Mexico|(5) 555-4729|(5) 555-3745|
|ANTON| Antonio Moreno Taquería|Antonio Moreno|Owner| Mataderos|2312 México D.F.|| 05023| Mexico|(5) 555-3932|
|...|...|...|...|...|...|...|...|...|...|...|

**Check** run a few queries on the provided database and try using the `LIKE` command.

<a name="guided-practice_4"></a>
## Guided Practice: DISTINCT and LIMIT Operators (10 mins)

The SELECT DISTINCT statement is used to return only distinct (different) values. In a table, a column may contain many duplicate values; and sometimes you only want to list the different (distinct) values.

### SQL SELECT DISTINCT Syntax
```sql
SELECT DISTINCT _column_name_,_column_name_  
FROM _table_name_;
```

#### SELECT DISTINCT Example

The following SQL statement selects only the distinct values from the "City"  columns from the "Customers" table:

```sql
SELECT DISTINCT "City" FROM customers;
```

| City|
|---|
|Leipzig|
|London|
|Nantes|
|...|

The `DISTINCT` operator is equivalent to the `numpy.unique` command.

### SQL SELECT LIMIT Syntax
Sometimes we may want to only retrieve a fixed number of records from the database. This is where the `LIMIT` operator comes in.

```sql
SELECT _column_name_,_column_name_  
FROM _table_name_
LIMIT _number_of_records;
```

#### SELECT LIMIT Example

The following SQL statement selects only the first 3 values from the "City"  columns from the "Customers" table:

```sql
SELECT DISTINCT "City"
FROM customers
LIMIT 3;
```

| City|
|---|
|Leipzig|
|London|
|Nantes|

The `LIMIT` operator is equivalent to the `pandas.Dataframe.head` command.

**Check** try running the same queries using `DISTINCT` and `LIMIT`

In [55]:
sql.read_sql("SELECT DISTINCT \"City\" FROM customers WHERE \"City\" LIKE '%L%' LIMIT 15 ;",con=conn)

City
0  Leipzig
1   Lander
2   London
3    Lille
4    Luleå
5   Lisboa
6     Lyon

<a name="guided-practice_5"></a>
## Guided Practice: GROUP BY Operator (10 mins)

A table may contain several records that have a common key. Consider for example the following `order_details` table:


| OrderID |  ProductID |  UnitPrice | Quantity | Discount |
| ----- |
|10248|11|14|12|0|
|10248|42|9.8|10|0|
|10248|72|34.8|5|0|
|10249|14|18.6|9|0|
|10249|51|42.4|40|0|
|10250|41|7.7|10|0|
|...|...|...|...|...|


The _GROUP BY_ statement is used in conjunction with the aggregate functions to group the result-set by one or more columns. For example we may want to know the total number of items purchased in each order.

#### SQL GROUP BY Syntax

The syntax to perform a _GROUP BY_ operation is the following:

```sql
SELECT column_name, aggregate_function(column_name)  
FROM table_name  
WHERE column_name operator value  
GROUP BY column_name;
```
We select the column we want to aggregate on and the function.


#### Example

In order to calculate the total amount of items in a specific order we can use the following query:

```sql
SELECT "OrderID", SUM("Quantity")
FROM order_details
GROUP BY "OrderID"
```

In [88]:
from pandasql import sqldf
from pandasql import load_meat, load_births
import psycopg2
from pandas.io import sql
login={ 'dbname': 'northwind',
      'user':'dsi_student',
      'port':5432,
      'host':'dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com',
      'password':'gastudents'}

conn=psycopg2.connect(**login)
c=conn.cursor()

In [94]:
q="SELECT \"OrderID\", COUNT(\"ProductID\") FROM order_details GROUP BY \"OrderID\" LIMIT 5"
sql.read_sql(q,con=conn)

OrderID  count
0    10248      3
1    10249      2
2    10250      3
3    10251      3
4    10252      3

In [63]:
meat

date    beef   veal    pork  lamb_and_mutton  broilers  \
0   1944-01-01   751.0   85.0  1280.0             89.0       NaN   
1   1944-02-01   713.0   77.0  1169.0             72.0       NaN   
2   1944-03-01   741.0   90.0  1128.0             75.0       NaN   
3   1944-04-01   650.0   89.0   978.0             66.0       NaN   
4   1944-05-01   681.0  106.0  1029.0             78.0       NaN   
5   1944-06-01   658.0  125.0   962.0             79.0       NaN   
6   1944-07-01   662.0  142.0   796.0             82.0       NaN   
7   1944-08-01   787.0  175.0   748.0             87.0       NaN   
8   1944-09-01   774.0  182.0   678.0             91.0       NaN   
9   1944-10-01   834.0  215.0   777.0            100.0       NaN   
10  1944-11-01   786.0  197.0   944.0             91.0       NaN   
11  1944-12-01   764.0  146.0  1013.0             91.0       NaN   
12  1945-01-01   820.0  119.0  1037.0            100.0       NaN   
13  1945-02-01   816.0   97.0   724.0             81.0       NaN   
14  1945-03-01   836.0  107.0   723.0             87.0       NaN   
15  1945-04-01   736.0   98.0   651.0             78.0       NaN   
16  1945-05-01   747.0  103.0   682.0             87.0       NaN   
17  1945-06-01   739.0  110.0   674.0             87.0       NaN   
18  1945-07-01   736.0  117.0   610.0             82.0       NaN   
19  1945-08-01   858.0  145.0   531.0             77.0       NaN   
20  1945-09-01   910.0  164.0   535.0             80.0       NaN   
21  1945-10-01  1022.0  201.0   628.0             95.0       NaN   
22  1945-11-01   933.0  173.0   945.0             86.0       NaN   
23  1945-12-01   783.0  118.0  1103.0             90.0       NaN   
24  1946-01-01   856.0  101.0  1051.0             74.0       NaN   
25  1946-02-01   827.0   85.0   966.0            109.0       NaN   
26  1946-03-01   796.0   92.0   780.0             97.0       NaN   
27  1946-04-01   734.0   91.0   846.0             83.0       NaN   
28  1946-05-01   605.0   83.0   810.0             65.0       NaN   
29  1946-06-01   461.0   78.0   505.0             75.0       NaN   
..         ...     ...    ...     ...              ...       ...   
797 2010-06-01  2320.0   10.7  1831.7             14.1    3139.7   
798 2010-07-01  2229.6   10.9  1702.2             12.8    3057.7   
799 2010-08-01  2286.6   11.2  1815.3             12.8    3200.0   
800 2010-09-01  2252.2   11.3  1883.5             13.1    3238.3   
801 2010-10-01  2234.9   11.4  2002.7             13.1    3183.7   
802 2010-11-01  2235.5   11.3  2068.0             14.4    3124.4   
803 2010-12-01  2270.9   11.7  2055.4             14.9    3175.0   
804 2011-01-01  2122.9   10.8  1896.2             11.2    3127.5   
805 2011-02-01  2020.4   10.5  1768.1             10.9    2852.4   
806 2011-03-01  2266.2   12.1  2054.4             14.1    3310.4   
807 2011-04-01  2052.5   10.2  1790.7             14.4    2973.5   
808 2011-05-01  2131.9   10.3  1759.7             12.9    3251.1   
809 2011-06-01  2375.0   11.2  1820.0             12.6    3284.6   
810 2011-07-01  2134.1   10.1  1637.1             10.9    3030.9   
811 2011-08-01  2386.9   11.4  1892.1             13.1    3330.3   
812 2011-09-01  2215.2   10.7  1954.4             11.8    3180.4   
813 2011-10-01  2215.1   10.6  2033.2             11.7    3096.8   
814 2011-11-01  2148.8   10.8  2086.7             12.6    2907.1   
815 2011-12-01  2126.3   10.8  2065.6             12.5    2855.8   
816 2012-01-01  2113.8   10.4  1987.0             12.1    3091.8   
817 2012-02-01  2009.0    9.8  1882.9             12.3    2954.0   
818 2012-03-01  2159.8   10.0  1987.9             14.2    3043.7   
819 2012-04-01  1990.6    9.9  1841.7             12.9    2993.8   
820 2012-05-01  2232.0   10.4  1926.8             13.6    3278.0   
821 2012-06-01  2252.1    8.9  1750.4             12.4    3105.3   
822 2012-07-01  2200.8    9.5  1721.8             12.5    3127.0   
823 2012-08-01  2367.5   10.1  1997.9             14.2    3317.4  

In [61]:
sql.read_sql("SELECT * FROM order_details GROUP BY \"OrderID\"", con=conn)

DatabaseError: Execution failed on sql: SELECT * FROM order_details GROUP BY "OrderID"
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

unable to rollback

<a name="guided-practice_6"></a>
## Guided Practice: HAVING Operator (10 mins)

The HAVING clause was added to SQL because the WHERE keyword could not be used with aggregate functions. Let's consider the previous example where we aggregated the `Quantity` of an order. What if we would like to filter the results for orders where the Maximum discout applied on any item is lower than 10%?

#### SQL HAVING Syntax

```sql
SELECT column_name, aggregate_function(column_name)  
FROM table_name  
WHERE column_name operator value  
GROUP BY column_name  
HAVING aggregate_function(column_name) operator value;
```

##### Example

```sql
SELECT "OrderID", SUM("Quantity"), MAX("Discount")
FROM order_details
GROUP BY "OrderID"
HAVING MAX("Discount") <= 0.1;
```

|OrderID|sum|max|
|---|
|10501|20|0|
|10295|4|0|
|10827|36|0|

> Check: can you use column aliases to make the results more readable?

<a name="ind-practice"></a>
## Independent Practice (10 minutes)

Now that you've learned a lot more commands, try combining them:

- retrieve the top 3 orders by number of items ordered
- retrieve the oldest 5 orders
- retrieve the 5 most recent orders. Nest that query into another one that ranks these 5 orders by shipping country (alphabetical order)
- try some queries on your own



> Answers:
```sql
SELECT "OrderID", SUM("Quantity") AS "Sum"
FROM order_details
GROUP BY "OrderID"
ORDER BY "Sum" DESC
LIMIT 3;
```
```sql
SELECT "OrderID", "OrderDate"
FROM orders
ORDER BY "OrderDate"
LIMIT 5;
```
```sql
SELECT * FROM (
	SELECT "OrderID", "OrderDate", "ShipCountry"
	FROM orders
	ORDER BY "OrderDate" DESC
	LIMIT 5
	) AS fivemostrecent
ORDER BY "ShipCountry";
```

<a name="conclusion"></a>
## Conclusion (5 mins)
In this lesson we have learned many more new commands to make our SQL queries more powerful.


In particular we have learned how to:
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`

> **Check** can you think of a few more business cases where these are useful?

***

### ADDITIONAL RESOURCES

- [The excellent pgexercises site](https://pgexercises.com/)
- [PostgreSQL Documentation](http://www.postgresql.org/docs/)
- [Sqlite Documentation](https://www.sqlite.org/docs.html)